<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Lei Sun" />

<meta name="date" content="2017-02-13" />

<title>Pipeline for null simulation</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">truncash</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/LSun/truncash">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Pipeline for null simulation</h1>
<h4 class="author"><em>Lei Sun</em></h4>
<h4 class="date"><em>2017-02-13</em></h4>

</div>


<p><strong>Last updated:</strong> 2018-05-15</p>
<strong>workflowr checks:</strong> <small>(Click a bullet for more information)</small>
<ul>
<li>
<details>
<p><summary> <strong style="color:blue;">✔</strong> <strong>R Markdown file:</strong> up-to-date </summary></p>
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</details>
</li>
<li>
<details>
<p><summary> <strong style="color:blue;">✔</strong> <strong>Environment:</strong> empty </summary></p>
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</details>
</li>
<li>
<details>
<p><summary> <strong style="color:blue;">✔</strong> <strong>Seed:</strong> <code>set.seed(12345)</code> </summary></p>
<p>The command <code>set.seed(12345)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</details>
</li>
<li>
<details>
<p><summary> <strong style="color:blue;">✔</strong> <strong>Session information:</strong> recorded </summary></p>
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</details>
</li>
<li>
<details>
<p><summary> <strong style="color:blue;">✔</strong> <strong>Repository version:</strong> <a href="https://github.com/LSun/truncash/tree/388e65e06000e313c170a82f3ed57346f6024897" target="_blank">388e65e</a> </summary></p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/.DS_Store
    Ignored:    analysis/BH_robustness_cache/
    Ignored:    analysis/FDR_Null_cache/
    Ignored:    analysis/FDR_null_betahat_cache/
    Ignored:    analysis/Rmosek_cache/
    Ignored:    analysis/StepDown_cache/
    Ignored:    analysis/alternative2_cache/
    Ignored:    analysis/alternative_cache/
    Ignored:    analysis/ash_gd_cache/
    Ignored:    analysis/average_cor_gtex_2_cache/
    Ignored:    analysis/average_cor_gtex_cache/
    Ignored:    analysis/brca_cache/
    Ignored:    analysis/cash_deconv_cache/
    Ignored:    analysis/cash_fdr_1_cache/
    Ignored:    analysis/cash_fdr_2_cache/
    Ignored:    analysis/cash_fdr_3_cache/
    Ignored:    analysis/cash_fdr_4_cache/
    Ignored:    analysis/cash_fdr_5_cache/
    Ignored:    analysis/cash_fdr_6_cache/
    Ignored:    analysis/cash_plots_cache/
    Ignored:    analysis/cash_sim_1_cache/
    Ignored:    analysis/cash_sim_2_cache/
    Ignored:    analysis/cash_sim_3_cache/
    Ignored:    analysis/cash_sim_4_cache/
    Ignored:    analysis/cash_sim_5_cache/
    Ignored:    analysis/cash_sim_6_cache/
    Ignored:    analysis/cash_sim_7_cache/
    Ignored:    analysis/correlated_z_2_cache/
    Ignored:    analysis/correlated_z_3_cache/
    Ignored:    analysis/correlated_z_cache/
    Ignored:    analysis/create_null_cache/
    Ignored:    analysis/cutoff_null_cache/
    Ignored:    analysis/design_matrix_2_cache/
    Ignored:    analysis/design_matrix_cache/
    Ignored:    analysis/diagnostic_ash_cache/
    Ignored:    analysis/diagnostic_correlated_z_2_cache/
    Ignored:    analysis/diagnostic_correlated_z_3_cache/
    Ignored:    analysis/diagnostic_correlated_z_cache/
    Ignored:    analysis/diagnostic_plot_2_cache/
    Ignored:    analysis/diagnostic_plot_cache/
    Ignored:    analysis/efron_leukemia_cache/
    Ignored:    analysis/fitting_normal_cache/
    Ignored:    analysis/gaussian_derivatives_2_cache/
    Ignored:    analysis/gaussian_derivatives_3_cache/
    Ignored:    analysis/gaussian_derivatives_4_cache/
    Ignored:    analysis/gaussian_derivatives_5_cache/
    Ignored:    analysis/gaussian_derivatives_cache/
    Ignored:    analysis/gd-ash_cache/
    Ignored:    analysis/gd_delta_cache/
    Ignored:    analysis/gd_lik_2_cache/
    Ignored:    analysis/gd_lik_cache/
    Ignored:    analysis/gd_w_cache/
    Ignored:    analysis/knockoff_10_cache/
    Ignored:    analysis/knockoff_2_cache/
    Ignored:    analysis/knockoff_3_cache/
    Ignored:    analysis/knockoff_4_cache/
    Ignored:    analysis/knockoff_5_cache/
    Ignored:    analysis/knockoff_6_cache/
    Ignored:    analysis/knockoff_7_cache/
    Ignored:    analysis/knockoff_8_cache/
    Ignored:    analysis/knockoff_9_cache/
    Ignored:    analysis/knockoff_cache/
    Ignored:    analysis/knockoff_var_cache/
    Ignored:    analysis/marginal_z_alternative_cache/
    Ignored:    analysis/marginal_z_cache/
    Ignored:    analysis/mosek_reg_2_cache/
    Ignored:    analysis/mosek_reg_4_cache/
    Ignored:    analysis/mosek_reg_5_cache/
    Ignored:    analysis/mosek_reg_6_cache/
    Ignored:    analysis/mosek_reg_cache/
    Ignored:    analysis/pihat0_null_cache/
    Ignored:    analysis/plot_diagnostic_cache/
    Ignored:    analysis/poster_obayes17_cache/
    Ignored:    analysis/real_data_simulation_2_cache/
    Ignored:    analysis/real_data_simulation_3_cache/
    Ignored:    analysis/real_data_simulation_4_cache/
    Ignored:    analysis/real_data_simulation_5_cache/
    Ignored:    analysis/real_data_simulation_cache/
    Ignored:    analysis/rmosek_primal_dual_2_cache/
    Ignored:    analysis/rmosek_primal_dual_cache/
    Ignored:    analysis/seqgendiff_cache/
    Ignored:    analysis/simulated_correlated_null_2_cache/
    Ignored:    analysis/simulated_correlated_null_3_cache/
    Ignored:    analysis/simulated_correlated_null_cache/
    Ignored:    analysis/simulation_real_se_2_cache/
    Ignored:    analysis/simulation_real_se_cache/
    Ignored:    analysis/smemo_2_cache/
    Ignored:    data/LSI/
    Ignored:    docs/.DS_Store
    Ignored:    docs/figure/.DS_Store
    Ignored:    output/fig/

</code></pre>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</details>
</li>
</ul>
<details>
<summary> <small><strong>Expand here to see past versions:</strong></small> </summary>
<ul>
<table style="border-collapse:separate; border-spacing:5px;">
<thead>
<tr>
<th style="text-align:left;">
File
</th>
<th style="text-align:left;">
Version
</th>
<th style="text-align:left;">
Author
</th>
<th style="text-align:left;">
Date
</th>
<th style="text-align:left;">
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
html
</td>
<td style="text-align:left;">
<a href="https://cdn.rawgit.com/LSun/truncash/e05bc836b3c74dc6ebca415afb5938675d6c5436/docs/nullpipeline.html" target="_blank">e05bc83</a>
</td>
<td style="text-align:left;">
LSun
</td>
<td style="text-align:left;">
2018-05-12
</td>
<td style="text-align:left;">
Update to 1.0
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/LSun/truncash/blob/cc0ab8379469bc3726f1508cd81e4ecd6fef1b1a/analysis/nullpipeline.Rmd" target="_blank">cc0ab83</a>
</td>
<td style="text-align:left;">
Lei Sun
</td>
<td style="text-align:left;">
2018-05-11
</td>
<td style="text-align:left;">
update
</td>
</tr>
<tr>
<td style="text-align:left;">
html
</td>
<td style="text-align:left;">
<a href="https://cdn.rawgit.com/LSun/truncash/0f36d998db26444c5dd01502ea1af7fbd1129b22/docs/nullpipeline.html" target="_blank">0f36d99</a>
</td>
<td style="text-align:left;">
LSun
</td>
<td style="text-align:left;">
2017-12-21
</td>
<td style="text-align:left;">
Build site.
</td>
</tr>
<tr>
<td style="text-align:left;">
html
</td>
<td style="text-align:left;">
<a href="https://cdn.rawgit.com/LSun/truncash/853a484bfacf347e109f6c8fb3ffaab5f4d6cc02/docs/nullpipeline.html" target="_blank">853a484</a>
</td>
<td style="text-align:left;">
LSun
</td>
<td style="text-align:left;">
2017-11-07
</td>
<td style="text-align:left;">
Build site.
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/LSun/truncash/blob/7db293de450c37c876711ea346e6a67ffb76b0f9/analysis/nullpipeline.Rmd" target="_blank">7db293d</a>
</td>
<td style="text-align:left;">
LSun
</td>
<td style="text-align:left;">
2017-02-14
</td>
<td style="text-align:left;">
index
</td>
</tr>
<tr>
<td style="text-align:left;">
html
</td>
<td style="text-align:left;">
<a href="https://cdn.rawgit.com/LSun/truncash/7db293de450c37c876711ea346e6a67ffb76b0f9/docs/nullpipeline.html" target="_blank">7db293d</a>
</td>
<td style="text-align:left;">
LSun
</td>
<td style="text-align:left;">
2017-02-14
</td>
<td style="text-align:left;">
index
</td>
</tr>
<tr>
<td style="text-align:left;">
html
</td>
<td style="text-align:left;">
<a href="https://cdn.rawgit.com/LSun/truncash/7693edf481bbc1680fad3a7e401259b57979806f/docs/nullpipeline.html" target="_blank">7693edf</a>
</td>
<td style="text-align:left;">
LSun
</td>
<td style="text-align:left;">
2017-02-14
</td>
<td style="text-align:left;">
pipeline
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/LSun/truncash/blob/ab7e0851e70e7cb174500c9adb0804f086a11758/analysis/nullpipeline.Rmd" target="_blank">ab7e085</a>
</td>
<td style="text-align:left;">
LSun
</td>
<td style="text-align:left;">
2017-02-14
</td>
<td style="text-align:left;">
Files commited by wflow_commit.
</td>
</tr>
<tr>
<td style="text-align:left;">
html
</td>
<td style="text-align:left;">
<a href="https://cdn.rawgit.com/LSun/truncash/3a12124c5fe872061be50ccf15ec7fa5c4c453be/docs/nullpipeline.html" target="_blank">3a12124</a>
</td>
<td style="text-align:left;">
LSun
</td>
<td style="text-align:left;">
2017-02-14
</td>
<td style="text-align:left;">
pipeline
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/LSun/truncash/blob/b68f251e41c935c37d7cd8dc85fbed335f1c580a/analysis/nullpipeline.Rmd" target="_blank">b68f251</a>
</td>
<td style="text-align:left;">
LSun
</td>
<td style="text-align:left;">
2017-02-14
</td>
<td style="text-align:left;">
Files commited by wflow_commit.
</td>
</tr>
<tr>
<td style="text-align:left;">
Rmd
</td>
<td style="text-align:left;">
<a href="https://github.com/LSun/truncash/blob/d623e66298e6a4b6595bbab18676b8843f59b7b0/analysis/nullpipeline.Rmd" target="_blank">d623e66</a>
</td>
<td style="text-align:left;">
LSun
</td>
<td style="text-align:left;">
2017-02-13
</td>
<td style="text-align:left;">
pipeline
</td>
</tr>
</tbody>
</table>
</ul>
</details>
<hr />
<p><strong>Last updated:</strong> 2018-05-15</p>
<p><strong>Code version:</strong> 388e65e</p>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>Up until now, we are using simulated correlated null data for exploration, and it would be helpful to examine and document the pipeline for generating these summary statistics, including <span class="math inline">\(\hat\beta\)</span>, <span class="math inline">\(\hat s\)</span>, <span class="math inline">\(z\)</span> score, <span class="math inline">\(t\)</span> statistics, <span class="math inline">\(p\)</span> value.</p>
</div>
<div id="pipeline" class="section level2">
<h2>Pipeline</h2>
<p>An expression matrix <span class="math inline">\(X_{G \times N}\)</span> will go through the following steps to get measurements of length <span class="math inline">\(g\)</span>: Choose the top <span class="math inline">\(g\)</span> expressed genes; randomly sample <span class="math inline">\(n\)</span> cases vs <span class="math inline">\(n\)</span> controls; calculate normalizing factor by <code>edgeR::calcNormFactors</code>; turn counts to <span class="math inline">\(\log_2\)</span>; calculate variance weight by <code>limma::voom</code>. We’ll go through this pipeline with a toy data set.</p>
<p><strong>Sometimes it’s very helpful to read the source code of these different functions.</strong></p>
<pre class="r"><code>library(edgeR)</code></pre>
<pre><code>Loading required package: limma</code></pre>
<pre class="r"><code>library(limma)</code></pre>
<div id="readgenerate-raw-counts" class="section level3">
<h3>read/generate raw counts</h3>
<p>Suppose the raw counts are stored in a <span class="math inline">\(10 \times 8\)</span> matrix with <span class="math inline">\(10\)</span> genes and <span class="math inline">\(8\)</span> samples, counts simulated from Possion.</p>
<pre class="r"><code>set.seed(777)
r = matrix(rpois(n = 80, lambda = 5), nrow = 10)
r</code></pre>
<pre><code>      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
 [1,]    6    6    4    7    1    6    6    2
 [2,]    5    6    6    6    4    4    3    7
 [3,]    4    5    5    7    5    2    7    9
 [4,]   12    5    3    4    4    3    0    4
 [5,]    6    7    1    4    3    5    8    4
 [6,]    1    2    7    4   11    3    2    8
 [7,]    4    4    5    6    8    2    2    4
 [8,]    3    8    5    4    8    9    3    6
 [9,]    9    4   11    6    8    4    9    5
[10,]    3    4    6    6    3    1    5    4</code></pre>
</div>
<div id="select-top-expressed-genes" class="section level3">
<h3>Select top expressed genes:</h3>
<ol style="list-style-type: decimal">
<li>convert each count to its lcpm.</li>
<li>select top expressed genes according their sum of lcpm.</li>
</ol>
<p>In this example we choose top <span class="math inline">\(6\)</span> genes from a total of <span class="math inline">\(10\)</span> genes.</p>
<pre class="r"><code>lcpm = function(r){R = colSums(r); t(log2(((t(r)+0.5)/(R+1))* 10^6))}
top_genes_index=function(g,X){return(order(rowSums(X),decreasing =TRUE)[1:g])}
Y=lcpm(r)
subset = top_genes_index(6,Y)
r = r[subset,]
r</code></pre>
<pre><code>     [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
[1,]    9    4   11    6    8    4    9    5
[2,]    3    8    5    4    8    9    3    6
[3,]    4    5    5    7    5    2    7    9
[4,]    5    6    6    6    4    4    3    7
[5,]    6    7    1    4    3    5    8    4
[6,]    6    6    4    7    1    6    6    2</code></pre>
</div>
<div id="randomly-sample-cases-and-controls" class="section level3">
<h3>randomly sample cases and controls</h3>
<p>In this example we use 2 cases vs 2 controls, and put labes on them by <code>condition</code>, based on which a design matrix is generated.</p>
<pre class="r"><code>counts = r[,sample(1:ncol(r),2*2)]
counts</code></pre>
<pre><code>     [,1] [,2] [,3] [,4]
[1,]    6    5   11    4
[2,]    4    6    5    9
[3,]    7    9    5    2
[4,]    6    7    6    4
[5,]    4    4    1    5
[6,]    7    2    4    6</code></pre>
<pre class="r"><code>condition = c(rep(0,2),rep(1,2))
condition</code></pre>
<pre><code>[1] 0 0 1 1</code></pre>
<pre class="r"><code>design = model.matrix(~condition)
design</code></pre>
<pre><code>  (Intercept) condition
1           1         0
2           1         0
3           1         1
4           1         1
attr(,&quot;assign&quot;)
[1] 0 1</code></pre>
</div>
<div id="calculate-normalizing-factors-for-each-sample-using-edgercalcnormfactors" class="section level3">
<h3>calculate normalizing factors for each sample using <code>edgeR::calcNormFactors</code></h3>
<p><span style="color:red">This might be potentially a problem, because the normalizing factors are only calculated based on top <span class="math inline">\(g\)</span> genes.</span> 1. A <code>DGEList</code> object is generated by <code>counts</code> and <code>condition</code>, using <code>edgeR::DGEList</code> 2. The normalizing factors are calculated by <code>edgeR::calcNormFactors</code>.</p>
<pre class="r"><code>DGEList_obj = edgeR::DGEList(counts=counts,group=condition)
DGEList_obj</code></pre>
<pre><code>An object of class &quot;DGEList&quot;
$counts
  Sample1 Sample2 Sample3 Sample4
1       6       5      11       4
2       4       6       5       9
3       7       9       5       2
4       6       7       6       4
5       4       4       1       5
6       7       2       4       6

$samples
        group lib.size norm.factors
Sample1     0       34            1
Sample2     0       33            1
Sample3     1       32            1
Sample4     1       30            1</code></pre>
<pre class="r"><code>dgecounts = edgeR::calcNormFactors(DGEList_obj)
dgecounts</code></pre>
<pre><code>An object of class &quot;DGEList&quot;
$counts
  Sample1 Sample2 Sample3 Sample4
1       6       5      11       4
2       4       6       5       9
3       7       9       5       2
4       6       7       6       4
5       4       4       1       5
6       7       2       4       6

$samples
        group lib.size norm.factors
Sample1     0       34    1.1201997
Sample2     0       33    0.9397760
Sample3     1       32    0.9051827
Sample4     1       30    1.0494070</code></pre>
</div>
<div id="calculate-weight-for-each-log_2-expression-using-limmavoom" class="section level3">
<h3>Calculate weight for each <span class="math inline">\(\log_2\)</span> expression using <code>limma::voom</code></h3>
<p><code>limma::voom</code> takes in counts and normalizing factors, and gives, among other things, modified library sizes, <span class="math inline">\(\log_2\)</span> expression, weights for each expression, all for Gaussian-based linear modeling.</p>
<p><span style="color:red">The peculiarity of <code>limma::voom</code> lies in the way it calculates the “average” log2-count for each gene.</span>: <code>sx &lt;- fit$Amean + mean(log2(lib.size + 1)) - log2(1e+06)</code>. Note that for each gene, the count as well as log2-count could vary wildly from sample to sample due to library size, sequencing depth, and / or experimental design, so the way to find an “average” is not self-evident.</p>
<pre class="r"><code>v = voom(dgecounts, design, plot=FALSE)
v</code></pre>
<pre><code>An object of class &quot;EList&quot;
$targets
        group lib.size norm.factors
Sample1     0 38.08679    1.1201997
Sample2     0 31.01261    0.9397760
Sample3     1 28.96585    0.9051827
Sample4     1 31.48221    1.0494070

$E
   Sample1  Sample2  Sample3  Sample4
1 17.34340 17.39043 18.54988 17.07992
2 16.81288 17.63144 17.48575 18.15792
3 17.54985 18.17893 17.48575 16.23192
4 17.34340 17.83789 17.72676 17.07992
5 16.81288 17.10093 15.61128 17.36942
6 17.54985 16.25293 17.19625 17.61043

$weights
         [,1]     [,2]     [,3]     [,4]
[1,] 1.849269 2.060690 1.849269 1.849269
[2,] 4.229762 2.147872 1.849269 1.849269
[3,] 1.849269 1.849269 1.260232 1.424488
[4,] 1.849269 4.502218 2.095619 2.171379
[5,] 2.133848 1.565478 1.260232 1.260232
[6,] 2.111233 1.462248 2.095619 2.171379

$design
  (Intercept) condition
1           1         0
2           1         0
3           1         1
4           1         1
attr(,&quot;assign&quot;)
[1] 0 1</code></pre>
<p>Here is what <code>limma::voom</code> does as in <a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29">Law et al 2013</a></p>
<p>voom: variance modeling at the observation-level</p>
<p>The limma-trend pipeline models the variance at the gene level. However, in RNA-seq applications, the count sizes may vary considerably from sample to sample for the same gene. Different samples may be sequenced to different depths, so different count sizes may be quite different even if the cpm values are the same. For this reason, we wish to model the mean-variance trend of the log-cpm values at the individual observation level, instead of applying a gene-level variability estimate to all observations from the same gene.</p>
<p>Our strategy is to estimate non-parametrically the mean-variance trend of the logged read counts and to use this mean-variance relationship to predict the variance of each log-cpm value. The predicted variance is then encapsulated as an inverse weight for the log-cpm value. When the weights are incorporated into a linear modeling procedure, the mean-variance relationship in the log-cpm values is effectively eliminated.</p>
<p><span style="color:red">A technical difficulty is that we want to predict the variances of individual observations although there is, by definition, no replication at the observational level from which variances could be estimated.</span> We work around this inconvenience by estimating the mean-variance trend at the gene level, then interpolating this trend to predict the variances of individual observations.</p>
<p>The algorithm proceeds as follows. First, gene-wise linear models are fitted to the normalized log-cpm values, taking into account the experimental design, treatment conditions, replicates and so on. This generates a residual standard deviation for each gene. A robust trend is then fitted to the residual standard deviations as a function of the average log-count for each gene.</p>
<p>Also available from the linear models is a fitted value for each log-cpm observation. Taking the library sizes into account, the fitted log-cpm for each observation is converted into a predicted count. The standard deviation trend is then interpolated to predict the standard deviation of each individual observation based on its predicted count size. Finally, the inverse squared predicted standard deviation for each observation becomes the weight for that observation.</p>
<p>The log-cpm values and associated weights are then input into limma’s standard differential expression pipeline. Most limma functions are designed to accept quantitative weights, providing the ability to perform microarray-like analyses while taking account of the mean-variance relationship of the log-cpm values at the observation level.</p>
</div>
<div id="perform-weighted-t-test-for-each-gene-and-obtain-primary-hatbeta-hat-s-t-d.f.-p-using-limmalmfit" class="section level3">
<h3>Perform weighted <span class="math inline">\(t\)</span>-test for each gene and obtain primary <span class="math inline">\(\hat\beta\)</span>, <span class="math inline">\(\hat s\)</span>, <span class="math inline">\(t\)</span>, d.f., <span class="math inline">\(p\)</span> using <code>limma::lmFit</code></h3>
<p>Taking in the <code>EList</code> from <code>limma::voom</code>, <code>limma::lmFit</code> produces <span class="math inline">\(g\)</span> <span class="math inline">\(t\)</span>-tests, giving vectors of <span class="math inline">\(\hat\beta\)</span>, <span class="math inline">\(\hat s\)</span>, degree of freedom.</p>
<p>Note that <code>$stdev.unscaled</code><span class="math inline">\(= (X^TWX)^{-1}\)</span>, <span class="math inline">\(W\)</span> being the weights calculated by <code>limma::voom</code>.</p>
<pre class="r"><code>lim = lmFit(v)
lim</code></pre>
<pre><code>An object of class &quot;MArrayLM&quot;
$coefficients
  (Intercept)  condition
1    17.36819  0.4467125
2    17.08856  0.7332766
3    17.86439 -1.0439088
4    17.69392 -0.2963208
5    16.93478 -0.4444242
6    17.01916  0.3878582

$stdev.unscaled
  (Intercept) condition
1   0.5057244 0.7253511
2   0.3959772 0.6535863
3   0.5199780 0.8017827
4   0.3967914 0.6259395
5   0.5199226 0.8167446
6   0.5289983 0.7170746

$sigma
[1] 1.0000276 0.8283184 0.8417619 0.6191636 1.0056442 0.9044428

$df.residual
[1] 2 2 2 2 2 2

$cov.coefficients
            (Intercept) condition
(Intercept)         0.5      -0.5
condition          -0.5       1.0

$pivot
[1] 1 2

$rank
[1] 2

$Amean
       1        2        3        4        5        6 
17.59091 17.52200 17.36161 17.49699 16.72363 17.15236 

$method
[1] &quot;ls&quot;

$design
  (Intercept) condition
1           1         0
2           1         0
3           1         1
4           1         1
attr(,&quot;assign&quot;)
[1] 0 1</code></pre>
<pre class="r"><code>betahat.lim = lim$coefficients[,2]
betahat.lim</code></pre>
<pre><code>         1          2          3          4          5          6 
 0.4467125  0.7332766 -1.0439088 -0.2963208 -0.4444242  0.3878582 </code></pre>
<pre class="r"><code>sebetahat.lim = lim$stdev.unscaled[,2]*lim$sigma
sebetahat.lim</code></pre>
<pre><code>        1         2         3         4         5         6 
0.7253712 0.5413776 0.6749101 0.3875590 0.8213545 0.6485530 </code></pre>
<pre class="r"><code>t.lim = betahat.lim / sebetahat.lim
df.lim = lim$df.residual
df.lim</code></pre>
<pre><code>[1] 2 2 2 2 2 2</code></pre>
</div>
<div id="do-variance-shrinkage-on-hat-s-and-obtain-moderated-t-d.f.-p-by-empirical-bayes-using-limmaebayes" class="section level3">
<h3>Do variance shrinkage on <span class="math inline">\(\hat s\)</span> and obtain moderated <span class="math inline">\(t\)</span>, d.f., <span class="math inline">\(p\)</span> by empirical Bayes using <code>limma::eBayes</code></h3>
<p><span style="color:red">Is it a variance shrinkage procedure?</span></p>
<p>Taking in summary statistics (<span class="math inline">\(\hat \beta\)</span>, <span class="math inline">\(\hat s\)</span>, d.f.) for each gene, <code>limma::eBayes</code> produces shrunk standard deviation for each gene, and thus, moderated <span class="math inline">\(t\)</span>-statistics and <span class="math inline">\(p\)</span>-values.</p>
<pre class="r"><code>lim.ebayes = ebayes(lim)
lim.ebayes</code></pre>
<pre><code>$df.prior
[1] Inf

$s2.prior
[1] 0.7679051

$s2.post
[1] 0.7679051 0.7679051 0.7679051 0.7679051 0.7679051 0.7679051

$t
  (Intercept)  condition
1    39.19105  0.7027908
2    49.24719  1.2802980
3    39.20572 -1.4857719
4    50.88712 -0.5402266
5    37.16953 -0.6209514
6    36.71385  0.6172411

$df.total
[1] 12 12 12 12 12 12

$p.value
   (Intercept) condition
1 4.912765e-14 0.4955946
2 3.220173e-15 0.2246354
3 4.890901e-14 0.1631317
4 2.177295e-15 0.5989240
5 9.230571e-14 0.5462455
6 1.069038e-13 0.5486090

$var.prior
[1] 20.83590882  0.01302244

$lods
  (Intercept) condition
1    751.8553 -4.601380
2   1196.5418 -4.585889
3    751.9251 -4.583232
4   1277.9980 -4.606774
5    675.1656 -4.601095
6    658.2650 -4.602920</code></pre>
<pre class="r"><code>sebetahat.ebayes = lim$stdev.unscaled[, 2] * sqrt(lim.ebayes$s2.post)
sebetahat.ebayes</code></pre>
<pre><code>        1         2         3         4         5         6 
0.6356266 0.5727390 0.7026037 0.5485120 0.7157149 0.6283739 </code></pre>
<pre class="r"><code>t.ebayes = betahat.lim / sebetahat.ebayes
t.ebayes</code></pre>
<pre><code>         1          2          3          4          5          6 
 0.7027908  1.2802980 -1.4857719 -0.5402266 -0.6209514  0.6172411 </code></pre>
<pre class="r"><code>lim.ebayes$t[, 2]</code></pre>
<pre><code>         1          2          3          4          5          6 
 0.7027908  1.2802980 -1.4857719 -0.5402266 -0.6209514  0.6172411 </code></pre>
<pre class="r"><code>df.ebayes = lim.ebayes$df.total
df.ebayes</code></pre>
<pre><code>[1] 12 12 12 12 12 12</code></pre>
<pre class="r"><code>p.ebayes = (1 - pt(abs(t.ebayes), df.ebayes)) * 2
p.ebayes</code></pre>
<pre><code>        1         2         3         4         5         6 
0.4955946 0.2246354 0.1631317 0.5989240 0.5462455 0.5486090 </code></pre>
<pre class="r"><code>lim.ebayes$p.value[, 2]</code></pre>
<pre><code>        1         2         3         4         5         6 
0.4955946 0.2246354 0.1631317 0.5989240 0.5462455 0.5486090 </code></pre>
</div>
<div id="obtain-z-scores-from-moderated-p-values-and-t-statistics" class="section level3">
<h3>obtain <span class="math inline">\(z\)</span> scores from moderated <span class="math inline">\(p\)</span>-values and <span class="math inline">\(t\)</span>-statistics</h3>
<pre class="r"><code>z = qnorm(1 - lim.ebayes$p[, 2] / 2) * sign(lim.ebayes$t[, 2])
z</code></pre>
<pre><code>         1          2          3          4          5          6 
 0.6814377  1.2142943 -1.3946160 -0.5259485 -0.6033956  0.5998458 </code></pre>
</div>
</div>
<div id="session-information" class="section level2">
<h2>Session Information</h2>
</div>
<div id="session-information-1" class="section level2">
<h2>Session information</h2>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.4.3 (2017-11-30)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS High Sierra 10.13.4

Matrix products: default
BLAS: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] edgeR_3.20.2 limma_3.34.4

loaded via a namespace (and not attached):
 [1] locfit_1.5-9.1    workflowr_1.0.1   Rcpp_0.12.16     
 [4] lattice_0.20-35   digest_0.6.15     rprojroot_1.3-2  
 [7] R.methodsS3_1.7.1 grid_3.4.3        backports_1.1.2  
[10] git2r_0.21.0      magrittr_1.5      evaluate_0.10.1  
[13] stringi_1.1.6     whisker_0.3-2     R.oo_1.21.0      
[16] R.utils_2.6.0     rmarkdown_1.9     tools_3.4.3      
[19] stringr_1.3.0     yaml_2.1.18       compiler_3.4.3   
[22] htmltools_0.3.6   knitr_1.20       </code></pre>
</div>

<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>

<hr>
<p>
  This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a>
  analysis was created with
  <a href="https://github.com/jdblischak/workflowr">workflowr</a> 1.0.1
</p>
<hr>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
